README
README
======

Kexx2 is a top down scrolling shoot'em up, similar to Raptor and Tyrian.

DISCLAIMER: Most of the code is from 2000-2001 and in an old C++ style (C with
classes). 

The project web page is at: http://www.haeggblad.com/kexx2
and the source code is hosted at: http://github.com/octol/kexx2

Installation from source
------------------------

If you've downloaded any of the source releases (e.g kexx2-1.0.9.tar.gz) then
running the following commands should suffice.

    ./configure                     (check for compilation dependencies)
    make                            (compile)
    make install                    (optional)

Running `make' should produce an executable `kexx2' in the src/ sub directory,
which `make install' then copies to the system (/usr/local/bin/). 

Requirements for building from source
-------------------------------------

To compile Kexx2 you will need a C++11 compiler and the following libraries:

- Boost::filesystem
- Boost::system
- Boost::program_options
- SDL 1.2
- SDL_mixer
- SDL_image


Installation from git repository
-------------------------------

When compiling from the git repository and additional step before `./configure'
is needed to produce the necessary build files. To generate these run

    autoreconf -ivf



NEWS
NEWS
====

2013-05-23  Jon Haggblad  <jon@haeggblad.com>

    As all bugs introduced due to the partial modernisation have now been
    fixed, I've decided to bump the version number to 1.1.0. Starting with this
    version I've also created pre-compiled binaries for Windows. These are
    statically linked so it should be just a matter of running the executable.

2013-05-14  Jon Haggblad  <jon@haeggblad.com>

    The loading of game resources has now been made robust, with an
    informative message if the data could not be found at runtime.
    Furthermore, the compile time path can be changed by specifying path
    using the --data argument at runtime. This directory will then be
    saved in the configuration file, which is now called
    $HOME/.kexx2.conf

2013-04-19  Jon Haggblad  <jon@haeggblad.com>

    Removal of most (all?) memory leaks by switching to the new
    reference counting objects in SDLc. Valgrind reports no leaks so
    far. The big class inheritance hierarchy has been broken up by
    making Object the new base class, instead of inheriting from
    sdlc::Sprite. 
    
    This release of Kexx2 goes together with SDLc 0.4.1.

2013-03-30  Jon Haggblad  <jon@haeggblad.com>

    Completed another cleanup. This time most C-style arrays was replaced by
    stdlib containers, raw pointers was replaced with smart pointers
    (std::unique_ptr, std::shared_ptr) and some sections was replaced with
    algorithms from <algorithm>.

2013-03-10  Jon Haggblad  <jon@haeggblad.com>

    As of today we are down to 10 extern system classes and 72 TODO
    items. This upcoming version (1.0.4) also marks a milestone in the
    sense that the entire source code a (rudimentary) cleanup. With this
    out of the way focus can now shift to do actual code changes.

2013-03-05  Jon Haggblad  <jon@haeggblad.com>

    Further updates. Moved to using sdlc namespace for all SDLc stuff now.
    Not yet migrated to using the meta header file SDLc.h

2013-02-28  Jon Haggblad  <jon@haeggblad.com>

    Refactoring is coming along nicely. So far it is mainly stylistic
    changes, such as setting default values in the class declarations,
    correct initialisation in constructors etc. Lots of variable
    renaming too, to make the code follow modern best practices.

    Hopefully when the first sweep is done, I can deal with more
    architectural issues, including removing the global system classes.
    Right now I'm down to 11 instances of the 'extern' keyword in the
    cpp files. Another goal is to move all raw SDL_ calls to the system
    classes.

2013-02-16  Jon Haggblad  <jon@haeggblad.com>

    After more than 10 years of being inactive, development has now
    resumed.  The goal is to modernise the code base and bring it up to
    modern C++11 standards.  As a first step I've updated the Autotools
    (Autoconf + Automake + Libtool) build system and I'm in the process
    of moving the source code from Sourceforge to Github. I have also
    decided to split out the classes used to wrap SDL to a separate
    library, SDLc ("SDL Classes"), also available on Github.

2003-06-10  Jon Haggblad  <jon@haeggblad.com>

    1.0.1b released. Includes just minor fixes (g++ 3.3 stuff among others).

2002-02-02  Jon Haggblad  <jon@haeggblad.com>

    1.0.0 finally released. Complete rewrite (again).

2001-06-10  Jon Haggblad  <jon@haeggblad.com>

    Complete rewrite with all the bells and whistles.
    Moved to UNIX type file structure.
    Trying to get the 1.0.0 out the door before Aug 20.

2001-06-03  Jon Haggblad  <jon@haeggblad.com>

    Had to comment out all the sound volume stuff for the 0.5.0 release. The 
    meter didn't show up for some weird reason and after I'd run the game
    once, windows would just die on me, almost every app would crash. After I 
    commented out it things returned to normal. 

2000-12-30  Jon Haggblad  <jon@haeggblad.com>

    Changed name of VisualC project files from "Project Kexx" to "Kexx2".
    Included a little "SDL Now" logo in the exit screen.
    The intro can now be bypassed with ESCAPE.
    The game now uses 10 pre-calculated big explosions instead of 1.

2000-12-26  Jon Haggblad  <jon@haeggblad.com>

    Ok 0.4.1 is out. 
    Options are now saved to data/kexx2.cfg when you exit the game.
    A new enemy type (EnemyRammer).
    Two player support.
    Getting a crash when I change the screen from fullscreen to windowed under 
    Win98: "Unable to init screen: DDraw created surface with wrong pitch".

2000-12-25  Jon Haggblad  <jon@haeggblad.com>

    Added new enemy class (EnemyRammer).
    Kexx2 now supports up to 20 levels.
    Added credits screen.
    Moved all the Visual C++ stuff to /visualc.
    Kexx2 is now under the GPL.

2000-12-24  Jon Haggblad  <jon@haeggblad.com>

    Restructured the data directory a little.
    Now reads from data/levels/levels.ini file to find out how many levels that 
    are available. Way better then the old "trial-and-error" way of doing things.

    Main::LoadLevel() needs to be rewritten soon, currently it works, but doesn't 
    scale very well with a lot of levels.



src/Menu.cpp
//    Kexx2
//    Copyright (C) 2001, 2013 Jon Häggblad
//
//    This file is part of Kexx2.
//
//    Kexx2 is free software: you can redistribute it and/or modify
//    it under the terms of the GNU General Public License as published by
//    the Free Software Foundation, either version 3 of the License, or
//    (at your option) any later version.
//
//    Kexx2 is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU General Public License for more details.
//
//    You should have received a copy of the GNU General Public License
//    along with Kexx2.  If not, see <http://www.gnu.org/licenses/>.

#include <iostream>

#include "Menu.h"

#include "config.h"

#include "SDLc/Timer.h"
#include "SDLc/Screen.h"
#include "SDLc/Mixer.h"
#include "SDLc/Font.h"
#include "SDLc/Input.h"

#include "PlayerState.h"
#include "Options.h"

// -----------------------------------------------------------------------------
// Construction/Destruction
// -----------------------------------------------------------------------------

Menu::Menu(sdlc::Mixer& mixer, Options& options) 
    : IGameState(EnvironmentType::menu), options_(options)
{
    load_data(mixer);
}

// -----------------------------------------------------------------------------
// Member Functions
// -----------------------------------------------------------------------------

void Menu::load_data(sdlc::Mixer& mixer)
{
    logo_.load(options_.data_path + "gfx/Menulogo.png");
    bg_music_.load(options_.data_path + "music/bgmusic1.xm");
    mixer.set_music_volume(mixer.music_volume());
    bg_music_.play(-1);
    selector_.gfx.load(options_.data_path + "gfx/EnemySideways.png");

    for (auto& b : bg_data_) {
        b.alloc(SCREEN_WIDTH, SCREEN_HEIGHT);
        precalc_bg_surface(b);
    }
}

void Menu::run_logic(sdlc::Input& input, sdlc::Timer& timer, sdlc::Mixer& mixer, 
                     PlayerState& player_state)
{
    // up/down
    if (input.key_pressed(SDLK_UP, sdlc::AutofireKeystate::off))
        selector_.pos--;
    else if (input.key_pressed(SDLK_DOWN, sdlc::AutofireKeystate::off))
        selector_.pos++;

    // limits
    if (selector_.pos < 0) 
        selector_.pos = 0;
    if (which_menu_ == State::root && selector_.pos > 2)
        selector_.pos = 2;
    else if (which_menu_ == State::options && selector_.pos > 3)
        selector_.pos = 3;

    // when return is pressed
    if (input.key_pressed(SDLK_RETURN, sdlc::AutofireKeystate::off)) {
        if (which_menu_ == State::root) 
            root_menu_logic(timer, mixer, player_state);
        else if (which_menu_ == State::options)
            options_menu_logic();
    }

    // exit screen
    if (which_menu_ == State::exit && timer.ticks() - exit_timer_ > 3000)
        done_ = true;

    // set option strings
    players_ = std::to_string(options_.num_of_players());
    display_ = options_.fullscreen() ? "fullscreen" : "windowed";
    fps_ = options_.fps_counter() ? "yes" : "no";
}

void Menu::draw(sdlc::Screen& screen, sdlc::Font& font)
{
    if (toggle_screen_) {
        screen.open(640, 480, 16, screen_type_);
        toggle_screen_ = false;
    }

    screen.blit(0, 0, bg_data_[rand() % 10]);

    if (which_menu_ != State::exit) {
        screen.blit(150, 60, logo_);
        screen.blit(230, 200 + selector_.pos * 30, selector_.gfx);
    }

    if (which_menu_ == State::root) {
        draw_root_screen(screen, font);
    } else if (which_menu_ == State::options) {
        draw_options_screen(screen, font);
    } else if (which_menu_ == State::exit) {
        draw_exit_screen(screen, font);
    } else {
        std::cout << "Menu::draw() undefined 'which_menu'" << std::endl;
    }

    const std::string v = VERSION;
    screen.print(5, 455, "v " + v, font);
}

// -----------------------------------------------------------------------------
// Private Functions
// -----------------------------------------------------------------------------

void Menu::root_menu_logic(sdlc::Timer& timer, sdlc::Mixer& mixer,
                           PlayerState& player_state)
{
    if (selector_.pos == 0) {
        player_state.kill_all();

        player_state.set_energy_max(1, 3);
        player_state.set_extra_weapon(1, "Rocket Weapon");
        player_state.set_extra_weapon_count(1, 10);

        if (options_.num_of_players() >= 2) {
            player_state.set_energy_max(2, 3);
            player_state.set_extra_weapon(2, "Rocket Weapon");
            player_state.set_extra_weapon_count(2, 10);
        }

        done_ = true;
    } else if (selector_.pos == 1) {
        which_menu_ = State::options;
        selector_.pos = 0;
    } else if (selector_.pos == 2) {
        player_state.kill_all();
        which_menu_ = State::exit;
        exit_timer_ = timer.ticks();
        mixer.fade_out_music(3000);
    }
}

void Menu::options_menu_logic()
{
    if (selector_.pos == 0) {
        options_.set_num_of_players(options_.num_of_players() + 1);
        if (options_.num_of_players() > NUM_OF_POSSIBLE_PLAYERS)
            options_.set_num_of_players(1);
    } else if (selector_.pos == 1) {
        options_.set_fullscreen(!options_.fullscreen());

#ifdef WIN32
        screen_type_ = options_.fullscreen() 
            //? SDL_HWSURFACE | SDL_DOUBLEBUF | SDL_FULLSCREEN 
            ? SDL_SWSURFACE | SDL_FULLSCREEN 
            : SDL_SWSURFACE;

        // no reinit of screen here for win32, they need to restart
        // the game for that
#endif
#ifndef WIN32
        screen_type_ = options_.fullscreen() 
            ? SDL_SWSURFACE | SDL_FULLSCREEN 
            : SDL_SWSURFACE;
        toggle_screen_ = true;
#endif
    } else if (selector_.pos == 2) {
        options_.set_fps_counter(!options_.fps_counter());
    } else if (selector_.pos == 3) {
        which_menu_ = State::root;
        selector_.pos = 1;
    }
}

void Menu::draw_root_screen(sdlc::Screen& screen, sdlc::Font& font)
{
    screen.print(270, 200, "start", font);
    screen.print(270, 230, "options", font);
    screen.print(270, 260, "exit", font);
}

void Menu::draw_options_screen(sdlc::Screen& screen, sdlc::Font& font)
{
    screen.print(270, 200, "players: " + players_, font);
    screen.print(270, 230, "display: " + display_, font);
    screen.print(270, 260, "show fps: " + fps_, font);
    screen.print(270, 300, "return", font);

#ifdef WIN32
    if (selector_.pos == 1)
        screen.print(230, 465, "changes requires a restart", 0, 0, 255);
#endif
}

void Menu::draw_exit_screen(sdlc::Screen& screen, sdlc::Font& font)
{
    static bool tmpB = false;
    if (tmpB == false)
        sdl_logo_.load(options_.data_path + "gfx/SDLnow.png");
    tmpB = true;
    screen.blit(640 - sdl_logo_.width(), 480 - sdl_logo_.height(), sdl_logo_);

    int x = 170, y = 100;
    screen.print(x, y, "programming by:", font);
    screen.print(x + 40, y + 30, "jon haggblad", font);

    screen.print(x, y + 70, "additional work by:", font);
    screen.print(x + 40, y + 100, "birger hedman", font);
    screen.print(x + 40, y + 120, "petter ulfberg", font);

    screen.print(x, y + 160, "special thanks to:", font);
    screen.print(x + 40, y + 190, "jesper rojestal", font);
    screen.print(x + 40, y + 210, "john degerman", font);
    screen.print(x + 40, y + 230, "markus haggblad", font);
}

void Menu::precalc_bg_surface(sdlc::Surface& surface)
{
    for (int iy = 0; iy < 40; iy++)
        for (int ix = 0; ix < 8; ix++)
            draw_group(ix * 80 + 5, iy * 12 + 2, surface);
}

void Menu::draw_group(int x, int y, sdlc::Surface& surface)
{
    for (int i = 0; i < 9; i++) 
        surface.print(x + (i * 8), y, std::to_string((rand() % 10)), 80, 80, 80);
}
